sharding/review/casper_v2.1.md

```diff
@@ -0,0 +1,380 @@
+* If `slot % EPOCH_LENGTH < EPOCH_LENGTH - 8`, let `start = shard_cutoffs[si]` and `end = shard_cutoffs[si+1]`. Otherwise, let `start = height_cutoffs[slot % EPOCH_LENGTH]` and `end = height_cutoffs[slot % EPOCH_LENGTH + 1]`
+* Verify that `len(attester_bitfield) == ceil_div8(end - start)`, where `ceil_div8 = (x + 7) // 8`. Verify that bits `end-start....` and higher, if present (ie. `end-start` is not a multiple of 8), are all zero
+* Take all indices `0 <= i < end-start` where the ith bit of `attester_bitfield` equals 1, take `current_epoch_shuffling[start+i]` as the indices of those validators, extract their pubkeys, and add them all together to generate the group pubkey.
+* Verify that the `checkpoint_hash` matches the `current_checkpoint`
+* Verify that the `aggregate_sig` verifies using the group pubkey and `hash(slot % EPOCH_LENGTH + parent + checkpoint_hash + shard_id + shard_block_hash)` as the message.
+* AND all indices taken above into the `attester_bitfield`. Add the `balance` of any newly added validators into the `total_attester_deposits`.
+
+Extend the list of `AttestationRecord` objects in the `active_state` , ordering the new additions by `shard_block_hash`.
+
+Verify that one of the `AttestationRecord` objects includes the first attester at the current height (ie. `current_epoch_shuffling[height_cutoffs[slot % EPOCH_LENGTH]]`); this attester can be considered to be the proposer of the block.
+
+### Epoch transitions
+
+When the current slot number passes into the next epoch, we execute an epoch transition, which breaks down into several parts. Note that all validator balance changes only take effect after all of these parts finish.
+
+#### Calculate rewards for FFG votes
+
+* If `total_attester_deposits * 3 >= total_deposits * 2`, set `crystallized_state.justified_epoch` to equal `crystallized_state.current_epoch` (note: this is still the previous epoch at this point in the computation). If this happens, and the justified epoch was previously `crystallized_state.current_epoch - 1`, set the `crystallized_state.finalized_epoch` to equal that value.
+* Compute the `online_reward` and `offline_penalty` based on the Casper FFG incentive and quadratic leak rules (not yet fully specified)
+* Add the `online_reward` to every validator who participated in the last epoch, and subtract the `offline_penalty` from everyone who did not
+* Set `attester_bitfield` to be all zeroes, with length `ceil_div8(len(active_validators))`. Set `total_attester_deposits` to 0
+
+### Calculate rewards for crosslinks
+
+Repeat for every shard S:
+
+* Take the set of `shard_block_hash` values that have been proposed for S in the attestations (call this set ROOTS)
+* Take the ROOT in ROOTS such that the total deposit size of the subset of validators that have voted for `shard_block_hash` ROOT and `shard_id` S is maximal; call this BESTROOT
```

Comment:
* Take `root` in `roots` such that the total deposit size of the subset of validators that have voted for `shard_block_hash` and shard `s` is maximal; call this `best_root`